compiler-plugin/src/main/kotlin/com/google/devtools/ksp/processing/impl/ResolverImpl.kt

```diff
@@ -1089,6 +1095,82 @@ class ResolverImpl(
     override fun getDeclarationsInSourceOrder(container: KSDeclarationContainer): Sequence<KSDeclaration> {
         return container.declarationsInSourceOrder
     }
+
+    @KspExperimental
+    override fun effectiveJavaModifiers(declaration: KSDeclaration): Set<Modifier> {
+        val modifiers = HashSet<Modifier>(declaration.modifiers.filter { it in javaModifiers })
+
+        // This is only needed by sources.
+        // PUBLIC, PRIVATE, PROTECTED are already handled in descriptor based impls.
+        fun addVisibilityModifiers() {
+            when {
+                declaration.isPublic() -> modifiers.add(Modifier.PUBLIC)
+                declaration.isPrivate() -> modifiers.add(Modifier.PRIVATE)
+                declaration.isProtected() -> modifiers.add(Modifier.PROTECTED)
+            }
+        }
+
+        when (declaration.origin) {
+            Origin.JAVA -> {
+                addVisibilityModifiers()
+                if (declaration is KSClassDeclaration && declaration.classKind == ClassKind.INTERFACE)
+                    modifiers.add(Modifier.ABSTRACT)
+            }
+            Origin.KOTLIN -> {
+                addVisibilityModifiers()
+                if (!declaration.isOpen())
+                    modifiers.add(Modifier.FINAL)
+                if (declaration.hasAnnotation("kotlin.jvm.JvmStatic"))
+                    modifiers.add(Modifier.JAVA_STATIC)
+                if (declaration.hasAnnotation("kotlin.jvm.JvmDefault"))
+                    modifiers.add(Modifier.JAVA_DEFAULT)
+                if (declaration.hasAnnotation("kotlin.jvm.JvmDefaultWithoutCompatibility"))
+                    modifiers.add(Modifier.JAVA_DEFAULT)
+                if (declaration.hasAnnotation("kotlin.jvm.Strictfp"))
+                    modifiers.add(Modifier.JAVA_STRICT)
+                if (declaration.hasAnnotation("kotlin.jvm.Synchronized"))
+                    modifiers.add(Modifier.JAVA_SYNCHRONIZED)
+                if (declaration.hasAnnotation("kotlin.jvm.Transient"))
+                    modifiers.add(Modifier.JAVA_TRANSIENT)
+                if (declaration.hasAnnotation("kotlin.jvm.Volatile"))
+                    modifiers.add(Modifier.JAVA_VOLATILE)
+                when (declaration) {
+                    is KSClassDeclaration -> {
+                        if (declaration.isCompanionObject)
+                            modifiers.add(Modifier.JAVA_STATIC)
+                        if (declaration.classKind == ClassKind.INTERFACE)
+                            modifiers.add(Modifier.ABSTRACT)
```

Comment:
Nit: feeling kind of tricky for this one. You can actually write `abstract interface` but it does not have any effects.

Reply:
This follows JVM [spec](https://docs.oracle.com/javase/specs/jvms/se8/html/jvms-4.html#jvms-4.1). Java interfaces have `abstract` **effectively**.

> If the ACC_INTERFACE flag is set, the ACC_ABSTRACT flag must also be set, and the ACC_FINAL, ACC_SUPER, and ACC_ENUM flags set must not be set.

Reply:
That's what I mean, interface is `abstract` by itself, adding a modifier seems a little bit unnecessary to me. Either way works for me.